nearcache.go

```diff
@@ -0,0 +1,1465 @@
+	defaultExpirationTaskInitialDelay = 5 * time.Second
+	// see: com.hazelcast.internal.nearcache.NearCache#DEFAULT_EXPIRATION_TASK_PERIOD_SECONDS
+	defaultExpirationTaskPeriod = 5 * time.Second
+)
+
+// serialization.Data is not hashable, so using this to store keys.
+type dataString string
+
+type nearCacheManager struct {
+	nearCaches   map[string]*nearCache
+	nearCachesMu *sync.RWMutex
+	ss           *serialization.Service
+	rt           *nearCacheReparingTask
+	lg           ilogger.LogAdaptor
+	doneCh       chan struct{}
+	state        int32
+}
+
+func newNearCacheManager(ic *client.Client, cfg *Config) *nearCacheManager {
+	doneCh := make(chan struct{})
+	cs := ic.ClusterService
+	is := ic.InvocationService
+	ss := ic.SerializationService
+	ps := ic.PartitionService
+	inf := ic.InvocationFactory
+	lg := ic.Logger
+	mf := newNearCacheInvalidationMetaDataFetcher(cs, is, inf, lg)
+	ris := cfg.Invalidation.ReconciliationIntervalSeconds()
+	mc := cfg.Invalidation.MaxToleratedMissCount()
+	rt := newNearCacheReparingTask(ris, mc, ss, ps, lg, mf, doneCh)
+	ncm := &nearCacheManager{
+		nearCaches:   map[string]*nearCache{},
+		nearCachesMu: &sync.RWMutex{},
+		ss:           ss,
+		rt:           rt,
+		lg:           lg,
+		doneCh:       doneCh,
+	}
+	return ncm
+}
+
+func (m *nearCacheManager) Stop() {
+	if atomic.CompareAndSwapInt32(&m.state, 0, 1) {
+		close(m.doneCh)
+	}
+}
+
+func (m *nearCacheManager) GetOrCreateNearCache(name string, cfg nearcache.Config) *nearCache {
+	m.nearCachesMu.RLock()
+	nc, ok := m.nearCaches[name]
+	m.nearCachesMu.RUnlock()
+	if ok {
+		return nc
+	}
+	m.nearCachesMu.Lock()
+	nc, ok = m.nearCaches[name]
+	if !ok {
+		delay := defaultExpirationTaskInitialDelay
+		period := defaultExpirationTaskPeriod
+		nc = newNearCache(&cfg, m.ss, delay, period, m.lg, m.doneCh)
+		m.nearCaches[name] = nc
+	}
+	m.nearCachesMu.Unlock()
+	return nc
+}
+
+type nearCacheUpdateSemantic int8
+
+const (
+	nearCacheUpdateSemanticReadUpdate nearCacheUpdateSemantic = iota
+	nearCacheUpdateSemanticWriteUpdate
+)
+
+type nearCache struct {
+	store                nearCacheRecordStore
+	cfg                  *nearcache.Config
+	logger               ilogger.LogAdaptor
+	doneCh               <-chan struct{}
+	expirationInProgress int32
+}
+
+func newNearCache(cfg *nearcache.Config, ss *serialization.Service, expirationInitialDelay, expirationPeriod time.Duration, logger ilogger.LogAdaptor, doneCh <-chan struct{}) *nearCache {
+	var rc nearCacheRecordValueConverter
+	var se nearCacheStorageEstimator
+	if cfg.InMemoryFormat == nearcache.InMemoryFormatBinary {
+		adapter := nearCacheDataStoreAdapter{ss: ss}
+		rc = adapter
+		se = adapter
+	} else {
+		adapter := nearCacheValueStoreAdapter{ss: ss}
+		rc = adapter
+		se = adapter
+	}
+	nc := &nearCache{
+		cfg:    cfg,
+		store:  newNearCacheRecordStore(cfg, ss, rc, se),
+		logger: logger,
+		doneCh: doneCh,
+	}
+	go nc.startExpirationTask(expirationInitialDelay, expirationPeriod)
```

Comment:
I think we should start this only if this condition is true: https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/DefaultNearCache.java#L103

Reply:
Fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/807/commits/adf231ea038a3c61a9ca428bc8a870b60458f291

Reply:
Really fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/807/commits/ebcf2b3b7fe69e1af93821996e0b1ea1d524c69d